### PR TITLE
Add support for tagging StructureFeatures

### PR DESCRIPTION
**Pull request by marchermans**
### Idea:
Support the grouping of structure features which get generated into a world together using tags.

### Reasoning:
It is generally useful to know the characteristics of structure features in a cross-mod compatible way.
This can be used by for example Entity AIs to discover structures where they do "research" or find artifacts. Or apply effects when the player is in some kind of structure.

An alternative for this tags would be for each mod to publish its own API and have other mods report this to said mod. But this ForgeRegistry based tag systems seems to be perfect for this solution, as it allows for the quick handling of defining a common set of structures which should be part of some list.

### Test mod example implementation:
The test mod just logs the positions of overworld temples (like desert pyramids and jungle) in the vicinity:
image

### CODE REVIEW

**Comment by Minecraftschurli** [comment]
wouldn't this also be handled by #8251 ?

**Comment by marchermans** [comment]
> wouldn't this also be handled by #8251 ?

No since this is an actual synced registry and can as such use normal types.

File: src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java [hunk 1]
```diff
@@ -234,4 +253,24 @@ public String getName()
             return "Mob Effect Tags";
         }
     }
+
+    public static class StructureFeatureTags extends ForgeRegistryTagsProvider<StructureFeature<?>> {
```
Comment (pupnewfster) [inline src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java]: Bracket should be on a new line

File: src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java [hunk 2]
```diff
@@ -115,6 +121,19 @@ private void leftClick(PlayerInteractEvent.LeftClickBlock event)
             if (blockEntity != null) logTagsIfPresent(blockEntity.getType().getTags());
             event.getEntityLiving().getActiveEffects().forEach((mobEffectInstance) -> logTagsIfPresent(mobEffectInstance.getEffect().getTags()));
         }
+
+        if (event.getPlayer().getLevel() instanceof ServerLevel serverLevel) {
```
Comment (pupnewfster) [inline src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java]: Bracket should be on a new line

File: src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java [hunk 3]
```diff
@@ -115,6 +121,19 @@ private void leftClick(PlayerInteractEvent.LeftClickBlock event)
...
+            OVERWORLD_TEMPLE.getValues().forEach(temple -> {
+                  final BlockPos found = serverLevel.findNearestMapFeature(temple, event.getPos(), 100, false);
+                  if (found != null) {
```
Comment (pupnewfster) [inline src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java]: Bracket should be on a new line

File: patches/minecraft/net/minecraft/world/level/levelgen/feature/StructureFeature.java.patch [hunk 4]
```diff
@@ -9,6 +9,30 @@
     public static final BiMap<String, StructureFeature<?>> f_67012_ = HashBiMap.create();
     private static final Map<StructureFeature<?>, GenerationStep.Decoration> f_67032_ = Maps.newHashMap();
     private static final Logger f_67033_ = LogManager.getLogger();
+@@ -79,6 +_,8 @@
+    public static final StructureFeature<JigsawConfiguration> f_67030_ = m_67089_("Bastion_Remnant", new BastionFeature(JigsawConfiguration.f_67756_), GenerationStep.Decoration.SURFACE_STRUCTURES);
+    public static final List<StructureFeature<?>> f_67031_ = ImmutableList.of(f_67013_, f_67028_, f_67029_, f_67022_);
+    public static final int f_160446_ = 8;
++
```
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/world/level/levelgen/feature/StructureFeature.java.patch]: This extra line break being added can be removed.

File: src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java [hunk 5]
```diff
@@ -234,4 +253,24 @@ public String getName()
...
+
+        public StructureFeatureTags(DataGenerator gen, @Nullable ExistingFileHelper existingFileHelper)
+        {
+            super(gen, ForgeRegistries.STRUCTURE_FEATURES, MODID, existingFileHelper);
+        }
+
+        @Override
+        public String getName()
+        {
+            return "StructureFeature Tags";
```
Comment (pupnewfster) [inline src/test/java/net/minecraftforge/debug/misc/CustomTagTypesTest.java]: I feel like there should be a space here `StructureFeature` ->  `Structure Feature`

**Review by pupnewfster — APPROVED** [review]

**Pull request merged**